### ISSUE TITLE
Feature extraction questions

### ISSUE THREAD

**Issue by cinjon** [wzmsltw/BSN-boundary-sensitive-network.pytorch#14]
Hi there, I am trying to train this on another dataset and am getting a bit stuck trying to figure out how exactly you extracted the features using TSN.

I am using the mmaction repository (which is what the authors of the TSN library you suggest ... suggest using) and the approach in that repository is to oversample by first computing the crops and flips and then run that through the model. 

I noticed in #3 that you said you don't remember if you used oversample or not. Has that changed by any chance? It would save me a lot of time if you can remember that.

Also, I noticed that the size of the features in the provided CSV were each of size 400. That seems small given that the TSN outputs features of size 1024 out of the box. Is there some other setting you used to get size 400 features?

Thanks for your help.

**Comment by cinjon** [comment]
Adding onto this a bit, let's say that we have a big video with 100,000 frames. Then l_s = l_v / sigma = 1e5 / 16 = 6250. Looking at the paper, it says that S = {s_n}_{n=1}^l_s, so this means that each snippets sequence has length 6250. It then says that the feature sequence is F = {f_{t_n}}_{n=1}^{l_s} so that also has length 6250. That seems way too big to me to be processing through the TSN.

Later on, you slide the windows of size 100 over the features. How does that work?

I suppose I'm a bit confused as to how this paper uses the features from TSN. Clarification on this would be very helpful.

**Comment by wzmsltw** [comment]
Hi @cinjon 
1. During feature extraction, I do not use oversampling and mulit-crops.
2. We get 400 dimension feature because I obtain feature from last layer of both rgb and flow-stream of TSN pretrained on activitynet, and activitynet have 200classes. Of cause you can use 1024 dimension layer to extract feature. I adopt last layer to save my disk space actually.
3. Why 6250 snippets is too big? 
4. In THUMOS, we slide window in feature sequence, I can send you codes of BSN in THUMOS if you give a email address.

**Comment by cinjon** [comment]
I just want to say I really appreciate you being so damn responsive here. What you've contributed here is a real boon to the community and I hope you understand that as well.

Some follow-ups:

1. Does that mean you just take the 224x224 center crop or not cropping at all? In either case, did you transform the image size first?
2. Oh I see, so does that mean you're taking the output of the layer in TSN that comes just before the softmax giving the probability scores for the classes? As a sidenote, it appears that the MMAction repo doesn't have the ActivityNet setup, so I'll reproduce it by taking the checkpoint from https://github.com/yjxiong/anet2016-cuhk and make a corresponding config using from https://github.com/yjxiong/temporal-segment-networks.

I am still confused though about how to proceed. I am trying this on my own dataset, so I don't think the BSN codes will be of much help. Here are some questions:

1. **What do you do with these snippets S of length 6250 to yield the features F of length 6250**? Each one of these snippets consists of an RGB x_{t_n} and a stack of five flows [o_{t_n-2}, o_{t_n-1}, o_{t_n}, o_{t_n+1}, o_{t_n+2}]. Do you just give the TSN the single RGB x_{t_n} to yield F_{S, t_n} and the five-tuple of Flows [o_{t_n-2}, o_{t_n-1}, o_{t_n}, o_{t_n+1}, o_{t_n+2}] to yield the F_{T, t_n}?

2. After yielding the features F, the windows then come into play. Here is how I understand that:

  A. Each window, indexed by k, is s.t. w_{100k, 100(k+1)} = {F_{100k, 100k+1, ..., 100(k+1)}, Y_{100k, 100k+1, ..., 100(k+1)}) where Y are the derived annotation bounds for the action regions that fall within the interval of time [100k, 100(k+1)). So if FPS = 1 and k = 1, then this would be the features from second 100 to second 200 and the annotations would be any annotation that has a start or end in this range. **What do you do with times that are outside of this range?** For example, does an annotation that was [95, 110] fall into both interval 0 as [95, 100] and interval 1 as [100, 110] or is there another scheme?

  B. The model then takes as input the entire F_w = F_{100k, 100k+1, ..., 100(k+1)}, i.e. 100 feature vectors, and produces the three probability sequences P_{S, w}, P_{E, W}, and P_{A, w}, each of length l_w = 100. Training happens on each of these intervals - i.e. the model does not know of anything beyond 100 length sequences. **Do you ever train on intervals that are shifted over, i.e. ones that start at frame 5 or 25 instead of 0?**

  C. **How do you then combine the results for each of the windows into the full result?** It appears that that happens in the PEM, but I'm a bit hazy on how this is happening.

**Comment by cinjon** [comment]
Ohhh, is F_{S, t_n} = TSN_rgb([x_{t_n - 8}, ..., x_{t_n}, ..., x_{t_n+7}], i.e. the sigma=16 frames? And then F_{T, t_n) = TSN_flow([o_{t_n-2}, o_{t_n-1}, o_{t_n}, o_{t_n+1}, o_{t_n+2}])?

**Comment by Zhuysheng** [comment]
@wzmsltw hi, Could you please send me codes of BSN in THUMOS, my email address is youknowzys@gmail.com

**Comment by dagnichz** [comment]
@wzmsltw  ,can you elaborate on the sliding window?
I see there is no explanation in the paper. and it is unclear how to process the TSN feature to always establish 100 len features?

**Comment by wzmsltw** [comment]
@Zhuysheng have sent

**Comment by wzmsltw** [comment]
@dagnichz For activitynet datset, plz refer https://github.com/wzmsltw/BSN-boundary-sensitive-network.pytorch/blob/master/data/activitynet_feature_cuhk/data_process.py  . For THUMOS-14 dataset, we adopt sliding window with length 100, plz give me your email then I can send you BSN-thumos codes.

**Comment by wzmsltw** [comment]
@cinjon 
1. I just resize image to 224x224 then extract feature, you can also try other configurations as well.
2. Yes
3. For each snippet which have 16 frames, I extract RGB feature using only the 8-th frame. and F_{T, t_n) = TSN_flow([o_{t_n-2}, o_{t_n-1}, o_{t_n}, o_{t_n+1}, o_{t_n+2}])
4. for details, you can give me a email then I can directly send you my codes of THUMOS.

**Comment by dagnichz** [comment]
my email is :dagnichz02@gmail.com.

thanks a lot!

**Comment by dagnichz** [comment]
> @dagnichz For activitynet datset, plz refer https://github.com/wzmsltw/BSN-boundary-sensitive-network.pytorch/blob/master/data/activitynet_feature_cuhk/data_process.py . For THUMOS-14 dataset, we adopt sliding window with length 100, plz give me your email then I can send you BSN-thumos codes.


my mail is :dagnichz02@gmail.com

**Comment by jojojo-h** [comment]
@wzmsltw hi,  Could you please send me codes of BSN in THUMOS14, my email address is wmy1245@gmail.com

**Comment by louisYen** [comment]
Could you send me BSN-thumos codes ?
My email address: m10502103@gmail.com
Thank you :)

**Comment by yechanp** [comment]
Can I get  BSN-thumos codes?
My address is : yechanp@gmail.com 
Thanks in advance!

**Comment by wzmsltw** [comment]
@yechanp @louisYen @jojojo-h @dagnichz have sent

**Comment by tobiascz** [comment]
I would also like to try the BSN-thumos code
mail: tobicgp@gmail.com
Thanks for your work!

**Comment by Qinying-Liu** [comment]
Can you share the BSN-thumos codes with me? My email is 1258168813@qq.com. Thank you!

**Comment by EscargotXJ** [comment]
@wzmsltw @Zhuysheng Would you share the BSN-thumos14 codes with me? My email is 1336742850@qq.com. Thank you!

**Comment by leemengxing** [comment]
@wzmsltw @yechanp @louisYen @jojojo-h @dagnichz hi everybody. I also want the BSN-thumos14 codes detail. my email is leemengxing93@gmail.com.Thank you!

**Comment by lakshaymehra** [comment]
@wzmsltw Can you please share the BSN in THUMOS code details at mehralakshay2@gmail.com ?Thanks!

**Comment by wzy1845** [comment]
> Hi there, I am trying to train this on another dataset and am getting a bit stuck trying to figure out how exactly you extracted the features using TSN.
> 
> I am using the mmaction repository (which is what the authors of the TSN library you suggest ... suggest using) and the approach in that repository is to oversample by first computing the crops and flips and then run that through the model.
> 
> I noticed in #3 that you said you don't remember if you used oversample or not. Has that changed by any chance? It would save me a lot of time if you can remember that.
> 
> Also, I noticed that the size of the features in the provided CSV were each of size 400. That seems small given that the TSN outputs features of size 1024 out of the box. Is there some other setting you used to get size 400 features?
> 
> Thanks for your help.

Hello! Have you successfully used TSN to extract features? Can you share the code of feature extraction? Thank you very much!!!

**Comment by dreamedrainbow** [comment]
> @yechanp @louisYen @jojojo-h @dagnichz have sent
Could you please send me codes of BSN in THUMOS14, my email address is rainbowdream1991@gmail.com. Thx a lot !

**Comment by WhoSJackie** [comment]
Could you please send me codes of BSN in THUMOS14, my email address is whosjackie@163.com. Thx a lot !

**Comment by NooneOnlyOne** [comment]
Would you share the BSN-thumos14 codes with me? My email is 15163881521@163.com Thanks

**Comment by lsj9731** [comment]
Could you send me BSN-thumos codes ?
My email address: not12m10@gmail.com
Thank you !!

**Comment by yangsuhui** [comment]
@wzmsltw Could you send me BSN-thumos14 codes ?
My email address: 857470845@qq.com
Thank you very much !!

**Comment by stuartchen1949** [comment]
Could you send me BSN-thumos14 codes ?
My email address: 2965914771@qq.com
 @wzmsltw Thank you very much !!

**Comment by OpenSorceYCW** [comment]
@wzmsltw Can you please share the BSN in THUMOS code details at 1083615449@qq.com ?Thanks!

**Comment by BigBuffa1o** [comment]
@wzmsltw Iam stuck at features truncate.Can you please share the BSN in THUMOS code details to me [1005617814@qq.com](mailto:1005617814@qq.com) ?Thanks!

**Comment by Rui-hue** [comment]
@wzmsltw Can you please share the BSN in THUMOS14 code details at [645249607@qq.com](mailto:645249607@qq.com) ?Thanks!
